modin/numpy/test/test_array.py

```diff
@@ -0,0 +1,762 @@
+    modin_result = modin_arr.prod(initial=0, where=False)
+    numpy_result = numpy_arr.prod(initial=0, where=False)
+    assert modin_result == numpy_result
+    modin_result = modin_arr.prod(keepdims=True)
+    numpy_result = numpy_arr.prod(keepdims=True)
+    assert modin_result.shape == numpy_result.shape
+    numpy.testing.assert_array_equal(modin_result._to_numpy(), numpy_result)
+    # Test 2D
+    numpy_arr = numpy.random.randint(-100, 100, size=(20, 20))
+    modin_arr = np.array(numpy_arr)
+    assert modin_arr.prod() == numpy_arr.prod()
+    modin_result = modin_arr.prod(axis=0)
+    numpy_result = numpy_arr.prod(axis=0)
+    assert modin_result.shape == numpy_result.shape
+    numpy.testing.assert_array_equal(modin_result._to_numpy(), numpy_result)
+    modin_result = modin_arr.prod(axis=0, keepdims=True)
+    numpy_result = numpy_arr.prod(axis=0, keepdims=True)
+    assert modin_result.shape == numpy_result.shape
+    numpy.testing.assert_array_equal(modin_result._to_numpy(), numpy_result)
+    modin_result = modin_arr.prod(axis=1)
+    numpy_result = numpy_arr.prod(axis=1)
+    assert modin_result.shape == numpy_result.shape
+    numpy.testing.assert_array_equal(modin_result._to_numpy(), numpy_result)
+    modin_result = modin_arr.prod(axis=1, keepdims=True)
+    numpy_result = numpy_arr.prod(axis=1, keepdims=True)
+    assert modin_result.shape == numpy_result.shape
+    numpy.testing.assert_array_equal(modin_result._to_numpy(), numpy_result)
+    modin_result = modin_arr.prod(initial=-200)
+    numpy_result = numpy_arr.prod(initial=-200)
+    assert modin_result == numpy_result
+    modin_result = modin_arr.prod(initial=0, where=False)
+    numpy_result = numpy_arr.prod(initial=0, where=False)
+    assert modin_result == numpy_result
+    with pytest.raises(ValueError):
+        modin_result = modin_arr.prod(out=modin_arr, keepdims=True)
+    modin_out = np.array([[1]])
+    numpy_out = modin_out._to_numpy()
+    modin_result = modin_arr.prod(out=modin_out, keepdims=True)
+    numpy_result = numpy_arr.prod(out=numpy_out, keepdims=True)
+    numpy.testing.assert_array_equal(modin_result._to_numpy(), numpy_result)
+    numpy.testing.assert_array_equal(modin_out._to_numpy(), numpy_out)
+    numpy_arr = numpy.random.randint(-100, 100, size=(20, 20))
+    modin_arr = np.array(numpy_arr)
+    modin_result = modin_arr.prod(axis=0, where=False, initial=4)
+    numpy_result = numpy_arr.prod(axis=0, where=False, initial=4)
+    numpy.testing.assert_array_equal(modin_result._to_numpy(), numpy_result)
+    numpy_out = numpy.ones(20)
+    modin_out = np.array(numpy_out)
+    modin_result = modin_arr.prod(axis=0, where=False, initial=4, out=modin_out)
+    numpy_result = numpy_arr.prod(axis=0, where=False, initial=4, out=numpy_out)
+    numpy.testing.assert_array_equal(modin_result._to_numpy(), numpy_result)
+    numpy.testing.assert_array_equal(modin_out._to_numpy(), numpy_out)
+    numpy_arr = numpy.random.randint(-5, 5, size=(20, 20))
```

Comment:
We have to change the range from [-100, 100] to [-5, 5] due to an interesting issue with overflow and dtypes that I'm not sure how we should handle - would love thoughts @devin-petersohn @dchigarev @YarShev @noloerino @vnlitvinov @jbrockmendel .

When we don't specify an `out` variable, the dtype of prod's result is np.int64, but when we specify an out variable, the dtype of prod's result is np.float64, and the arrays are different:

```python
In [1]: import numpy as np

In [2]: arr = np.random.randint(-100, 100, size=(20, 20))

In [3]: arr_out = np.ones(20)

In [4]: arr
Out[4]:
array([[   4,   47,  -16,   99,   39,  -90,   34,   94,   23,   42,   59,
          37,  -72, -100,  -85,  -84,  -15,   42,  -31,  -14],
       [ -33,  -62,  -33,   70,   -4,   16,  -12,  -95,  -61,   20,  -24,
          38,    2,   34,   44,  -83,  -27,   96,  -80,   77],
       [  30,    3,   65,  -71,   89,   32,  -88,  -73,  -16,   78,  -48,
         -25,   66,   18,   61,   -3,  -25,   65,  -24,  -40],
       [  42,   91,   33,   89,    4,  -75,  -14,   51,  -38,   28,   51,
          50,   21,  -83,  -90,   79,   51,  -24,    1,  -31],
       [  62,   29,   88,   39,   35,  -13,    5,  -13,   56,  -55,  -47,
         -12,   91,   72,   16,  -77,   17,  -16,   91,    8],
       [  55,   51,   63,   71,   93,   32,   13,   53,   10,  -28,  -23,
         -21,  -16,   -1,  -57,   27,  -82,  -54,   76,  -23],
       [  -3,   42,  -68,    8,  -68,   69,  -13,  -63,   21,  -14,  -65,
         -33,  -92,   46,   82,  -44,  -49,   31,  -58,   15],
       [ -78,  -66,  -78,  -57,   21,  -32,   31,   77,   66,   74,   32,
          -9,   89,   -8,   72,    9,   24,  -81,  -69,  -57],
       [  73,  -34,  -99,  -55,   77,   97,  -47,   -6,  -17,  -97,   55,
         -47,  -27,   16,  -34,  -31,   60,   75,  -31,  -77],
       [  28,   93,   70,  -34,   16,   97, -100,  -71,   -6,    7,  -85,
         -36,  -85,   58,   63,  -24,   59,   51,  -53,   12],
       [ -46,  -85,  -71,  -27,   58,  -59,   99,   68,  -78,   59,  -62,
          43,  -84,   22,   12,   21,  -32,  -39,   -4,   53],
       [ -41,   -1,    6,   23,   96,   96,   40,  -26,   22,   76,  -66,
          57,   59,   78,    2,   70,   46,  -62,   63,  -24],
       [ -53,  -58,  -66,  -32,  -41,   94,  -63,   87,  -56,   39,  -18,
          82,  -29,   78,   46,   87,   67,  -19,  -80,  -51],
       [ -99,    5,  -92,  -32,  -69,  -74,   50,   28,  -44,   13,  -42,
         -27,  -88,  -38,    7,  -50,  -66,   10,  -84,   57],
       [  63,  -64,   94,   19,    9,  -47,   78,  -87,   44,  -19,  -73,
          25,   24,   96,   72,   32,   17,   63,  -71,  -66],
       [  78,   78,   76,  -23,  -68,   80,   58,  -39,   46,   14,   38,
          90,   49,  -73,   73,   52,   21,  -87, -100,   89],
       [  49,   87,  -49,  -94,  -41,   -3,   49,  -69,  -19,   97,  -36,
         -62,   -9,  -81,   45,   90,   34,   56,   70,   94],
       [ -37,   94,  -25,   35,   74,   54,  -66,   -6,    0,  -58,   99,
         -16,  -77,   86,   71,   91,   95,  -22,    0,   42],
       [  35,  -37,  -67,  -42,   93,   87,  -49,  -54,  -11,   17,  -59,
          30,   68,  -92,   36,  -96,  -22,   77,   38,  -46],
       [ -31,  -72,   82,    4,  -25,   38,   44,  -72,   68,   66,  -12,
           8,   28,  -66,   38,  -17,   76,   56,  -66,    7]])

In [5]: arr_out
Out[5]:
array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1.])

In [6]: arr.prod(axis=0, initial=4)
Out[6]:
array([ 2091070475769245696,  9122834620979347456, -2313693447860518912,
       -6986630960492052480, -7019186078701584384, -2279247918881832960,
        4273937260288671744, -8551503026969591808,                    0,
       -8847250197584150528,  1372478237506011136, -2455170360243912704,
       -6397790424269848576,  4147333037207584768,  8496979910332514304,
       -1303936011327242240, -1330519711226003456,  6572564016281944064,
                          0,  5809144060652552192])

In [7]: arr.prod(axis=0, initial=4, out=arr_out)
Out[7]:
array([-1.03793712e+32, -1.10588070e+32, -1.14812384e+35,  6.35256562e+31,
       -8.65060786e+31,  9.47890390e+33, -9.74011828e+31, -4.45383972e+33,
        0.00000000e+00,  3.03451960e+31,  3.15602280e+33,  2.58784961e+30,
        4.28673938e+32, -7.53221323e+32,  1.68168256e+32,  5.75774327e+32,
        1.50429190e+32, -2.79290502e+33, -0.00000000e+00,  2.14612714e+31])

In [8]: arr_out
Out[8]:
array([-1.03793712e+32, -1.10588070e+32, -1.14812384e+35,  6.35256562e+31,
       -8.65060786e+31,  9.47890390e+33, -9.74011828e+31, -4.45383972e+33,
        0.00000000e+00,  3.03451960e+31,  3.15602280e+33,  2.58784961e+30,
        4.28673938e+32, -7.53221323e+32,  1.68168256e+32,  5.75774327e+32,
        1.50429190e+32, -2.79290502e+33, -0.00000000e+00,  2.14612714e+31])

In [9]: arr_out.dtype
Out[9]: dtype('float64')

In [10]: arr.prod(axis=0, initial=4).dtype
Out[10]: dtype('int64')
```
For now, I've hopefully bypassed the issue by picking a range that shouldn't overflow, but would love to hear folks' thoughts on how to deal with this!

Reply:
I ended up casting to the output dtype before performing computation, which allows us to replicate NumPy's behavior!

Reply:
If I'm reading this correctly, the output in [6] does not match the output in [8].  Setting aside dtype differences, shouldn't the values match?

Reply:
I thought they should too at first, but if my hunch about overflow is correct, it kind of makes sense? the float64 dtype has a broader range, so we get the actual answer, while the int64 range is much lower, so we get some random number thanks to overflow?